Fix Clang-trunk warnings about special members deprecated in C++20.

Quuxplusone opened the pull request:
For example:

    include/flatbuffers/reflection.h:365:8: error: definition of implicit copy
          constructor for 'pointer_inside_vector<flatbuffers::Table, unsigned char>'
          is deprecated because it has a user-declared copy assignment operator
          [-Werror,-Wdeprecated-copy]
      void operator=(const pointer_inside_vector &piv);
           ^

It's unclear why the old code wanted to declare a public `operator=`
without defining it; that just seems like a misunderstanding of the C++03 idiom
for deleting a member function. And anyway, we don't *want* to delete the
assignment operator; these are polymorphic types that do not follow value
semantics and nobody should ever be trying to copy them. So the simplest fix
is just to go back to the Rule of Zero: remove the declaration of `operator=`
and let the compiler do what it wanted to do originally anyway.
"The best code is no code."

Fixes #5649.

Quuxplusone commented:
@vglavnyy: I don't see any obvious way to add you as an "Assignee" and/or "Reviewer", but, please take a look.

vglavnyy commented on include/flatbuffers/reflection.h:
```diff
@@ -362,7 +362,6 @@ template<typename T, typename U> class pointer_inside_vector {
         reinterpret_cast<uint8_t *>(flatbuffers::vector_data(vec_)) + offset_);
   }
   T *operator->() const { return operator*(); }
-  void operator=(const pointer_inside_vector &piv);
```
I agree this declaration is not needed.
`std::vector<U> &vec_;` is non-copyable and non-movable.

Can you remove a similar declaration in `reflection.cpp`?
https://github.com/google/flatbuffers/blob/c4231c3cb99159978ca5f0669a07ef1aa02b88e3/src/reflection.cpp#L299

vglavnyy commented on include/flatbuffers/reflection_generated.h:
```diff
@@ -161,7 +161,6 @@ struct TypeBuilder {
         : fbb_(_fbb) {
     start_ = fbb_.StartTable();
   }
-  TypeBuilder &operator=(const TypeBuilder &);
```
I agree, `flatbuffers::FlatBufferBuilder &fbb_;` blocks copy and move even if there was an idea to make `&operator=` private.

Quuxplusone commented:
Updated (including your suggested fix to reflection.cpp).

I think that it would probably be useful to investigate and eliminate all uses of `FLATBUFFERS_DELETE_FUNC`; it seems like it's been applied pretty haphazardly and again isn't really doing anything except "protecting" against a very specific logic bug that nobody _should_ ever be writing. However, I'm not terribly motivated to do that patch myself because Clang isn't (yet? :) ) complaining about anything except the stuff I fixed in this PR.

aardappel commented:
LGTM
